numpy/_core/src/multiarray/static_string.c

```diff
@@ -0,0 +1,628 @@
+#define NPY_STRING_ON_HEAP 0x10      // 0001 0000
+#define NPY_STRING_MEDIUM 0x08       // 0000 1000
+#define NPY_STRING_FLAG_MASK 0xF8    // 1111 1000
+
+// short string sizes fit in a 4-bit integer
+#define NPY_SHORT_STRING_SIZE_MASK 0x0F  // 0000 1111
+#define NPY_SHORT_STRING_MAX_SIZE \
+    (sizeof(npy_static_string) - 1)      // 15 or 7 depending on arch
+#define NPY_MEDIUM_STRING_MAX_SIZE 0xFF  // 256
+
+// Since this has no flags set, technically this is a heap-allocated string
+// with size zero. Practically, that doesn't matter because we always do size
+// checks before accessing heap data, but that may be confusing. The nice part
+// of this choice is a calloc'd array buffer (e.g. from np.empty) is filled
+// with empty elements for free
+const _npy_static_string_u empty_string_u = {
+        .direct_buffer = {.size_and_flags = 0, .buf = {0}}};
+
+#define VSTRING_FLAGS(string) \
+    string->direct_buffer.size_and_flags & ~NPY_SHORT_STRING_SIZE_MASK;
+#define HIGH_BYTE_MASK ((size_t)0XFF << 8 * (sizeof(size_t) - 1))
+#define VSTRING_SIZE(string) (string->vstring.size_and_flags & ~HIGH_BYTE_MASK)
+
+typedef struct npy_string_arena {
+    size_t cursor;
+    size_t size;
+    char *buffer;
+} npy_string_arena;
+
+struct npy_string_allocator {
+    npy_string_malloc_func malloc;
+    npy_string_free_func free;
+    npy_string_realloc_func realloc;
+    npy_string_arena arena;
+};
+
+void
+set_vstring_size(_npy_static_string_u *str, size_t size)
+{
+    unsigned char current_flags = str->direct_buffer.size_and_flags;
+    str->vstring.size_and_flags = size;
+    str->direct_buffer.size_and_flags = current_flags;
+}
+
+char *
+vstring_buffer(npy_string_arena *arena, _npy_static_string_u *string)
+{
+    char flags = VSTRING_FLAGS(string);
+    if (flags & NPY_STRING_ON_HEAP) {
+        return (char *)string->vstring.offset;
+    }
+    if (arena->buffer == NULL) {
+        return NULL;
+    }
+    return (char *)((size_t)arena->buffer + string->vstring.offset);
+}
+
+#define ARENA_EXPAND_FACTOR 1.25
+
+char *
+arena_malloc(npy_string_arena *arena, npy_string_realloc_func r, size_t size)
+{
+    // one extra size_t to store the size of the allocation
+    size_t string_storage_size;
+    if (size <= NPY_MEDIUM_STRING_MAX_SIZE) {
+        string_storage_size = size + sizeof(unsigned char);
+    }
+    else {
+        string_storage_size = size + sizeof(size_t);
+    }
+    if ((arena->size - arena->cursor) <= string_storage_size) {
+        // realloc the buffer so there is enough room
+        // first guess is to double the size of the buffer
+        size_t newsize;
+        if (arena->size == 0) {
+            newsize = string_storage_size;
+        }
+        else if (((ARENA_EXPAND_FACTOR * arena->size) - arena->cursor) >
+                 string_storage_size) {
+            newsize = ARENA_EXPAND_FACTOR * arena->size;
+        }
+        else {
+            newsize = arena->size + string_storage_size;
+        }
```

Comment:
That does look nice but I don't want to grow the buffer by a power of two. I have it set to 1.25 right now to avoid over-allocating too much, since right now we don't have a way to realloc to fix exactly once we're done filling the arena. I also don't want to limit it to grow by 1 MiB at a time (in principle someone could put the contents of the library of congress in a string...). So not quite a perfect match. I could refactor both usages to use a common implementation but I'd prefer not to right now.

Reply:
It is 25%.  I used it as `newsize = arena->cursor + string_storage_size`.  If `newsize < arena->size` then grow `newsize` (not arena size).

Reply:
Ah, sorry, I misread the intent of the `min_grow` parameter as an expansion factor. That's what I get for not reading the docstrings...